PR: https://github.com/odoo/odoo/pull/169955

From: e0f281433eec3f34d6f720656ead51e417679cd1
From: Einar Andersen (eian)
Date: 2024-09-12 00:18:30

Breaking data model changes scores: del:5 + add:0, change matches:
-    description = fields.Html("Description")
-    partner_mobile = fields.Char(related="candidate_id.partner_mobile", readonly=False)
-    partner_mobile_sanitized = fields.Char(related="candidate_id.partner_mobile_sanitized")
-    partner_mobile = fields.Char(
-    partner_mobile_sanitized = fields.Char(

Total Changes: 190

[IMP] hr_recruitment: UX

A rework of the UX/UI for parts of the recruitment app which includes
* Simplify the UI for creating job position for a more seamless UX.
* Improved and added missing placeholders.
* Fixed spelling errors.
* A new toggle widget with a clickable label!.
* Improved naming of Reports, Refusal Reasons.
* Adds the possibility of loading demo recruitment data.
* Simplified and restructured the create applicant form view for better UX.
* Added the ability to update the email of a candidate (in case the candidate did not have one set) when refusing the candidate.
* Removed the ability to create applications and candidates through calendar view.
* Improved naming for reporting views and filters.
* Graph and Pivot view now display the same information.
* Improved naming for the settings to better display the intent.
* Improved demo data to make skill search work.
* Offers for refused candidates will now have the refusal reason set to "Others" if it exists.
* A confirmation dialog for when archiving a job position that has at least one application.
* Only department manager and recruiter are now the default followers of job positions.
* "CV" renamed to "Résumé" in settings.
* It is now possible to send and email or SMS to candidates from the list view, just like on applications.
* Job positions now have properties.

task-3850740

closes odoo/odoo#169955

Related: odoo/enterprise#64982
Related: odoo/upgrade#6314
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_recruitment/models/hr_applicant.py
+++ b/addons/hr_recruitment/models/hr_applicant.py
@@ -35,7 +35,6 @@ class Applicant(models.Model):
     _track_duration_field = 'stage_id'
 
     active = fields.Boolean("Active", default=True, help="If the active field is set to false, it will allow you to hide the case without removing it.", index=True)
-    description = fields.Html("Description")
 
     candidate_id = fields.Many2one('hr.candidate', required=True, index=True)
     partner_id = fields.Many2one(related="candidate_id.partner_id")
@@ -44,8 +43,6 @@ class Applicant(models.Model):
     email_normalized = fields.Char(related="candidate_id.email_normalized")
     partner_phone = fields.Char(related="candidate_id.partner_phone", readonly=False)
     partner_phone_sanitized = fields.Char(related="candidate_id.partner_phone_sanitized")
-    partner_mobile = fields.Char(related="candidate_id.partner_mobile", readonly=False)
-    partner_mobile_sanitized = fields.Char(related="candidate_id.partner_mobile_sanitized")
     linkedin_profile = fields.Char(related="candidate_id.linkedin_profile", readonly=False)
     type_id = fields.Many2one(related="candidate_id.type_id", readonly=False)
     availability = fields.Date(related="candidate_id.availability", readonly=False)
@@ -74,11 +71,11 @@ class Applicant(models.Model):
     date_open = fields.Datetime("Assigned", readonly=True)
     date_last_stage_update = fields.Datetime("Last Stage Update", index=True, default=fields.Datetime.now)
     priority = fields.Selection(AVAILABLE_PRIORITIES, "Evaluation", default='0')
-    job_id = fields.Many2one('hr.job', "Applied Job", domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", tracking=True, index=True)
+    job_id = fields.Many2one('hr.job', "Job Position", domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", tracking=True, index=True)
     salary_proposed_extra = fields.Char("Proposed Salary Extra", help="Salary Proposed by the Organisation, extra advantages", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
     salary_expected_extra = fields.Char("Expected Salary Extra", help="Salary Expected by Applicant, extra advantages", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
-    salary_proposed = fields.Float("Proposed Salary", aggregator="avg", help="Salary Proposed by the Organisation", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
-    salary_expected = fields.Float("Expected Salary", aggregator="avg", help="Salary Expected by Applicant", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
+    salary_proposed = fields.Float("Proposed", aggregator="avg", help="Salary Proposed by the Organisation", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
+    salary_expected = fields.Float("Expected", aggregator="avg", help="Salary Expected by Applicant", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
     department_id = fields.Many2one(
         'hr.department', "Department", compute='_compute_department', store=True, readonly=False,
         domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", tracking=True)
@@ -102,7 +99,7 @@ class Applicant(models.Model):
     meeting_display_date = fields.Date(compute='_compute_meeting_display')
     # UTMs - enforcing the fact that we want to 'set null' when relation is unlinked
     campaign_id = fields.Many2one(ondelete='set null')
-    medium_id = fields.Many2one(ondelete='set null')
+    medium_id = fields.Many2one(ondelete='set null', help="This displays how the applicant has reached out, e.g. via Email, LinkedIn, Website, etc.")
     source_id = fields.Many2one(ondelete='set null')
     interviewer_ids = fields.Many2many('res.users', 'hr_applicant_res_users_interviewers_rel',
         string='Interviewers', index=True, tracking=True,
@@ -115,6 +112,7 @@ class Applicant(models.Model):
     ], compute="_compute_application_status", search="_search_application_status")
     other_applications_count = fields.Integer(compute='_compute_other_applications_count', compute_sudo=True)
     applicant_properties = fields.Properties('Properties', definition='job_id.applicant_properties_definition', copy=True)
+    applicant_notes = fields.Html()
     refuse_date = fields.Datetime('Refuse Date')
 
     def init(self):
@@ -292,7 +290,7 @@ class Applicant(models.Model):
     def _phone_get_number_fields(self):
         """ This method returns the fields to use to find the number to use to
         send an SMS on a record. """
-        return ['partner_mobile', 'partner_phone']
+        return ['partner_phone']
 
     @api.depends('stage_id.hired_stage')
     def _compute_date_closed(self):
@@ -311,19 +309,24 @@ class Applicant(models.Model):
                 vals['email_from'] = vals['email_from'].strip()
         applicants = super().create(vals_list)
         applicants.sudo().interviewer_ids._create_recruitment_interviewers()
-        # Record creation through calendar, creates the calendar event directly, it will also create the activity.
-        if 'default_activity_date_deadline' in self.env.context:
-            deadline = fields.Datetime.to_datetime(self.env.context.get('default_activity_date_deadline'))
+
+        if (applicants.interviewer_ids.partner_id - self.env.user.partner_id):
             for applicant in applicants:
-                partners = applicant.partner_id | applicant.user_id.partner_id | applicant.department_id.manager_id.user_id.partner_id
-                self.env['calendar.event'].sudo().with_context(default_applicant_id=applicant.id).create({
-                    'applicant_id': applicant.id,
-                    'partner_ids': [(6, 0, partners.ids)],
-                    'user_id': self.env.uid,
-                    'name': applicant.partner_name,
-                    'start': deadline,
-                    'stop': deadline + relativedelta(minutes=30),
-                })
+                interviewers_to_notify = applicant.interviewer_ids.partner_id - self.env.user.partner_id
+                notification_subject = _("You have been assigned as an interviewer for %s", applicant.display_name)
+                notification_body = _("You have been assigned as an interviewer for the Applicant %s", applicant.partner_name)
+                applicant.message_notify(
+                    res_id=applicant.id,
+                    model=applicant._name,
+                    partner_ids=interviewers_to_notify.ids,
+                    author_id=self.env.user.partner_id.id,
+                    email_from=self.env.user.email_formatted,
+                    subject=notification_subject,
+                    body=notification_body,
+                    email_layout_xmlid="mail.mail_notification_layout",
+                    record_name=applicant.display_name,
+                    model_description="Applicant",
+                )
         # Copy CV from candidate to applicant at record creation
         attachments_result = self.env['ir.attachment'].read_group([
             ('res_id', 'in', applicants.candidate_id.ids),
@@ -359,6 +362,24 @@ class Applicant(models.Model):
             interviewers_to_clean = old_interviewers - self.interviewer_ids
             interviewers_to_clean._remove_recruitment_interviewers()
             self.sudo().interviewer_ids._create_recruitment_interviewers()
+            self.message_unsubscribe(partner_ids=interviewers_to_clean.partner_id.ids)
+
+            new_interviewers = self.interviewer_ids - old_interviewers - self.env.user
+            if new_interviewers:
+                notification_subject = _("You have been assigned as an interviewer for %s", self.display_name)
+                notification_body = _("You have been assigned as an interviewer for the Applicant %s", self.partner_name)
+                self.message_notify(
+                    res_id=self.id,
+                    model=self._name,
+                    partner_ids=new_interviewers.partner_id.ids,
+                    author_id=self.env.user.partner_id.id,
+                    email_from=self.env.user.email_formatted,
+                    subject=notification_subject,
+                    body=notification_body,
+                    email_layout_xmlid="mail.mail_notification_layout",
+                    record_name=self.display_name,
+                    model_description="Applicant",
+                )
         if vals.get('date_closed'):
             for applicant in self:
                 if applicant.job_id.date_to:
@@ -375,10 +396,8 @@ class Applicant(models.Model):
 
         nocontent_body = Markup("""
 <p class="o_view_nocontent_smiling_face">%(help_title)s</p>
-<p class="mb-0">%(para_1)s<br/>%(para_2)s</p>""") % {
+""") % {
             'help_title': _("No application found. Let's create one !"),
-            'para_1': _('People can also apply by email to save time.'),
-            'para_2': _("You can search into attachment's content, like resumes, with the searchbar."),
         }
 
         if hr_job:
@@ -389,12 +408,12 @@ class Applicant(models.Model):
             'para_1': match[1],
             'para_2': match[2],
             'para_3': match[3],
-            'link': f'/odoo/recruitement/{hr_job.id}',
+            'link': f'/odoo/recruitment/{hr_job.id}',
         }
 
         if hr_job.alias_email:
             nocontent_body += Markup('<p class="o_copy_paste_email oe_view_nocontent_alias">%(helper_email)s <a href="mailto:%(email)s">%(email)s</a></p>') % {
-                'helper_email': _("Create new applications by sending an email to"),
+                'helper_email': _("Try creating an application by sending an email to"),
                 'email': hr_job.alias_email,
             }
 

--- a/addons/hr_recruitment/models/hr_candidate.py
+++ b/addons/hr_recruitment/models/hr_candidate.py
@@ -49,18 +49,6 @@ class HrCandidate(models.Model):
         compute='_compute_partner_phone_sanitized',
         store=True,
         index='btree_not_null')
-    partner_mobile = fields.Char(
-        string="Mobile",
-        size=32,
-        compute='_compute_partner_phone_email',
-        inverse='_inverse_partner_email',
-        store=True,
-        index='btree_not_null')
-    partner_mobile_sanitized = fields.Char(
-        string='Sanitized Mobile Number',
-        compute='_compute_partner_mobile_sanitized',
-        store=True,
-        index='btree_not_null')
     linkedin_profile = fields.Char('LinkedIn Profile')
     type_id = fields.Many2one('hr.recruitment.degree', "Degree")
     availability = fields.Date("Availability", help="The date at which the applicant will be available to start working", tracking=True)
@@ -96,7 +84,7 @@ class HrCandidate(models.Model):
     def init(self):
         self.env.cr.execute("""
             CREATE INDEX IF NOT EXISTS hr_candidate_email_partner_phone_mobile
-            ON hr_candidate(email_normalized, partner_mobile_sanitized, partner_phone_sanitized);
+            ON hr_candidate(email_normalized, partner_phone_sanitized);
         """)
 
     @api.depends('partner_name')
@@ -109,11 +97,6 @@ class HrCandidate(models.Model):
         for candidate in self:
             candidate.partner_phone_sanitized = candidate._phone_format(fname='partner_phone') or candidate.partner_phone
 
-    @api.depends('partner_mobile')
-    def _compute_partner_mobile_sanitized(self):
-        for candidate in self:
-            candidate.partner_mobile_sanitized = candidate._phone_format(fname='partner_mobile') or candidate.partner_mobile
-
     @api.depends('partner_id')
     def _compute_partner_phone_email(self):
         for candidate in self:
@@ -122,8 +105,6 @@ class HrCandidate(models.Model):
             candidate.email_from = candidate.partner_id.email
             if not candidate.partner_phone:
                 candidate.partner_phone = candidate.partner_id.phone
-            if not candidate.partner_mobile:
-                candidate.partner_mobile = candidate.partner_id.mobile
 
     def _inverse_partner_email(self):
         for candidate in self:
@@ -139,12 +120,10 @@ class HrCandidate(models.Model):
                 # change email on a partner will trigger other heavy code, so avoid to change the email when
                 # it is the same. E.g. "email@example.com" vs "My Email" <email@example.com>""
                 candidate.partner_id.email = candidate.email_from
-            if candidate.partner_mobile:
-                candidate.partner_id.mobile = candidate.partner_mobile
             if candidate.partner_phone:
                 candidate.partner_id.phone = candidate.partner_phone
 
-    @api.depends('email_from', 'partner_mobile_sanitized', 'partner_phone_sanitized')
+    @api.depends('email_from', 'partner_phone_sanitized')
     def _compute_similar_candidates_count(self):
         """
             The field similar_candidates_count is only used on the form view.
@@ -159,7 +138,7 @@ class HrCandidate(models.Model):
                 else:
                     candidate.similar_candidates_count = 0
             return
-        self.flush_recordset(['email_normalized', 'partner_phone_sanitized', 'partner_mobile_sanitized'])
+        self.flush_recordset(['email_normalized', 'partner_phone_sanitized'])
         self.env.cr.execute("""
             SELECT
                 id,
@@ -168,9 +147,6 @@ class HrCandidate(models.Model):
                     FROM hr_candidate AS sub
                     WHERE c.id != sub.id
                      AND ((coalesce(c.email_normalized, '') <> '' AND sub.email_normalized = c.email_normalized)
-                       OR (coalesce(c.partner_mobile_sanitized, '') <> '' AND c.partner_mobile_sanitized = sub.partner_mobile_sanitized)
-                       OR (coalesce(c.partner_mobile_sanitized, '') <> '' AND c.partner_mobile_sanitized = sub.partner_phone_sanitized)
-                       OR (coalesce(c.partner_phone_sanitized, '') <> '' AND c.partner_phone_sanitized = sub.partner_mobile_sanitized)
                        OR (coalesce(c.partner_phone_sanitized, '') <> '' AND c.partner_phone_sanitized = sub.partner_phone_sanitized))
                 ) AS similar_candidates
             FROM hr_candidate AS c
@@ -184,7 +160,7 @@ class HrCandidate(models.Model):
     def _get_similar_candidates_domain(self):
         """
             This method returns a domain for the applicants whitch match with the
-            current candidate according to email_from, partner_phone or partner_mobile.
+            current candidate according to email_from, partner_phone.
             Thus, search on the domain will return the current candidate as well if any of
             the following fields are filled.
         """
@@ -195,9 +171,7 @@ class HrCandidate(models.Model):
         if self.email_normalized:
             domain = expression.OR([domain, [('email_normalized', '=', self.email_normalized)]])
         if self.partner_phone_sanitized:
-            domain = expression.OR([domain, ['|', ('partner_phone_sanitized', '=', self.partner_phone_sanitized), ('partner_mobile_sanitized', '=', self.partner_phone_sanitized)]])
-        if self.partner_mobile_sanitized:
-            domain = expression.OR([domain, ['|', ('partner_mobile_sanitized', '=', self.partner_mobile_sanitized), ('partner_phone_sanitized', '=', self.partner_mobile_sanitized)]])
+            domain = expression.OR([domain, [('partner_phone_sanitized', '=', self.partner_phone_sanitized)]])
         return domain
 
     def _compute_attachment_count(self):
@@ -386,7 +360,7 @@ class HrCandidate(models.Model):
             'lang': address_sudo.lang,
             'address_id': self.company_id.partner_id.id,
             'candidate_id': self.ids,
-            'phone': self.partner_phone or self.partner_mobile
+            'phone': self.partner_phone
         }
 
     def _check_interviewer_access(self):
@@ -411,3 +385,15 @@ class HrCandidate(models.Model):
             'search_view_id': self.env.ref('hr_recruitment.ir_attachment_view_search_inherit_hr_recruitment').ids,
             'domain': [('res_model', '=', 'hr.candidate'), ('res_id', 'in', self.ids)],
         }
+
+    def action_send_email(self):
+        return {
+            'name': _('Send Email'),
+            'type': 'ir.actions.act_window',
+            'target': 'new',
+            'view_mode': 'form',
+            'res_model': 'candidate.send.mail',
+            'context': {
+                'default_candidate_ids': self.ids,
+            }
+        }

--- a/addons/hr_recruitment/models/hr_job.py
+++ b/addons/hr_recruitment/models/hr_job.py
@@ -3,8 +3,11 @@
 import ast
 from collections import defaultdict
 
+from dateutil.relativedelta import relativedelta
 from odoo import api, fields, models, SUPERUSER_ID, _
 from odoo.tools import SQL
+from odoo.tools.convert import convert_file
+
 
 class Job(models.Model):
     _name = "hr.job"
@@ -45,7 +48,7 @@ class Job(models.Model):
         readonly=True, store=True)
     user_id = fields.Many2one('res.users', "Recruiter",
         domain="[('share', '=', False), ('company_ids', 'in', company_id)]", default=lambda self: self.env.user,
-        tracking=True, help="The Recruiter will be the default value for all Applicants Recruiter's field in this job \
+        tracking=True, help="The Recruiter will be the default value for all Applicants in this job \
             position. The Recruiter is automatically added to all meetings with the Applicant.")
     document_ids = fields.One2many('ir.attachment', compute='_compute_document_ids', string="Documents", readonly=True)
     documents_count = fields.Integer(compute='_compute_document_ids', string="Document Count")
@@ -62,6 +65,8 @@ class Job(models.Model):
     activities_overdue = fields.Integer(compute='_compute_activities')
     activities_today = fields.Integer(compute='_compute_activities')
 
+    job_properties = fields.Properties('Properties', definition='company_id.job_properties_definition')
+
     applicant_properties_definition = fields.PropertiesDefinition('Applicant Properties')
     no_of_hired_employee = fields.Integer(
         compute='_compute_no_of_hired_employee',
@@ -249,7 +254,7 @@ class Job(models.Model):
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
-            vals['favorite_user_ids'] = vals.get('favorite_user_ids', []) + [(4, self.env.uid)]
+            vals["favorite_user_ids"] = vals.get("favorite_user_ids", [])
         jobs = super().create(vals_list)
         utm_linkedin = self.env.ref("utm.utm_source_linkedin", raise_if_not_found=False)
         if utm_linkedin:
@@ -259,10 +264,21 @@ class Job(models.Model):
             } for job in jobs]
             self.env['hr.recruitment.source'].create(source_vals)
         jobs.sudo().interviewer_ids._create_recruitment_interviewers()
+        # Automatically subscribe the department manager and the recruiter to a job position.
+        for job in jobs:
+            job.message_subscribe(
+                job.manager_id._get_related_partners().ids + job.user_id.partner_id.ids
+            )
+
         return jobs
 
     def write(self, vals):
         old_interviewers = self.interviewer_ids
+        old_managers = {}
+        old_recruiters = {}
+        for job in self:
+            old_managers[job] = job.manager_id
+            old_recruiters[job] = job.user_id
         if 'active' in vals and not vals['active']:
             self.application_ids.active = False
         res = super().write(vals)
@@ -271,6 +287,35 @@ class Job(models.Model):
             interviewers_to_clean._remove_recruitment_interviewers()
             self.sudo().interviewer_ids._create_recruitment_interviewers()
 
+        # Subscribe the department manager if the department has changed
+        if "department_id" in vals:
+            for job in self:
+                to_unsubscribe = [
+                    partner
+                    for partner in old_managers[job]._get_related_partners().ids
+                    if partner not in job.user_id.partner_id.ids
+                ]
+                job.message_unsubscribe(to_unsubscribe)
+                job.message_subscribe(job.manager_id._get_related_partners().ids)
+
+        # Subscribe the recruiter if it has changed.
+        if "user_id" in vals:
+            for job in self:
+                to_unsubscribe = [
+                    partner
+                    for partner in old_recruiters[job].partner_id.ids
+                    if partner not in job.manager_id._get_related_partners().ids
+                ]
+                job.message_unsubscribe(to_unsubscribe)
+                job.message_subscribe(job.user_id.partner_id.ids)
+
+        # Update the availability on all hired candidates if the mission end date is changed
+        if "date_to" in vals:
+            for job in self:
+                hired_candidates = job.application_ids.filtered(lambda a: a.application_status == 'hired')
+                for candidate in hired_candidates:
+                    candidate.availability = job.date_to + relativedelta(days=1)
+
         # Since the alias is created upon record creation, the default values do not reflect the current values unless
         # specifically rewritten
         # List of fields to keep synched with the alias
@@ -350,7 +395,24 @@ class Job(models.Model):
             'name': _('Job'),
             'res_model': 'hr.job',
             'res_id': self.id,
-            'views': [(form_view.id, 'form'),],
+            'views': [(form_view.id, 'form')],
             'type': 'ir.actions.act_window',
             'target': 'inline'
         }
+
+    @api.model
+    def _action_load_recruitment_scenario(self):
+
+        convert_file(
+            self.env,
+            "hr_recruitment",
+            "data/scenarios/hr_recruitment_scenario.xml",
+            None,
+            mode="init",
+            kind="data",
+        )
+
+        return {
+            "type": "ir.actions.client",
+            "tag": "reload",
+        }

--- a/addons/hr_recruitment/models/hr_recruitment_stage.py
+++ b/addons/hr_recruitment/models/hr_recruitment_stage.py
@@ -14,7 +14,7 @@ class RecruitmentStage(models.Model):
         "Sequence", default=10)
     job_ids = fields.Many2many(
         'hr.job', string='Job Specific',
-        help='Specific jobs that uses this stage. Other jobs will not use this stage.')
+        help='Specific jobs that use this stage. Other jobs will not use this stage.')
     requirements = fields.Text("Requirements")
     template_id = fields.Many2one(
         'mail.template', "Email Template",

--- a/addons/hr_recruitment/models/res_company.py
+++ b/addons/hr_recruitment/models/res_company.py
@@ -7,3 +7,4 @@ class ResCompany(models.Model):
     _inherit = "res.company"
 
     candidate_properties_definition = fields.PropertiesDefinition('Candidate Properties')
+    job_properties_definition = fields.PropertiesDefinition("Job Properties")
